### PR TITLE
Documenting Port API's

### PR DESCRIPTION
**Pull request by takekoputa**
This is not finished yet.

Documented:
- Port / PortProxy / TranslatingPortProxy
- MasterPort / SlavePort
- ExternalMaster / ExternalSlave
- <s>RubyPortProxy</s>

A few notes:
- I don't know users need access to Ruby beyond the RubyPortProxy. I don't know whether RubyPort itself should be a part of the API's. If this is the case then should we also documenting PortProxy etc.

### CODE REVIEW

File: src/mem/external_master.hh [hunk 1]
```diff
@@ -115,19 +124,36 @@ class ExternalMaster : public SimObject
...
+     *
+     * @ingroup api_port
+     */
     static void registerHandler(const std::string &handler_name,
         Handler *handler);
 
+    /**
+      * @ingroup api_port
+      */
     void init() override;
```
Comment (powerjg) [inline src/mem/external_master.hh:152]: Not part of this API, it's part of the SimObject API

File: src/mem/external_master.hh [hunk 2]
```diff
@@ -70,12 +70,17 @@ class ExternalMaster : public SimObject
         ExternalMaster &owner;
 
       public:
+        /**
+         * @ingroup api_port
```
Comment (powerjg) [inline src/mem/external_master.hh:74]: External ports are different from the "port" API. Maybe this should be "external_api"?

File: src/mem/external_master.hh [hunk 3]
```diff
@@ -115,19 +124,36 @@ class ExternalMaster : public SimObject
     static std::map<std::string, Handler *> portHandlers;
 
   public:
+    /**
+      * @ingroup api_port
+      */
     ExternalMaster(ExternalMasterParams *params);
```
Comment (powerjg) [inline src/mem/external_master.hh:130]: I think constructors are part of the SimObject API... but I could be wrong. Thoughts @BobbyRBruce?

File: src/mem/external_master.hh [hunk 4]
```diff
@@ -115,19 +124,36 @@ class ExternalMaster : public SimObject
...
 
+    /**
+      * @ingroup api_port
+      */
     void init() override;
 
+    /**
+      * @ingroup api_port
+      */
     const MasterID masterId;
```
Comment (powerjg) [inline src/mem/external_master.hh:157]: Should this be public? Is it used publicly anywhere?

File: src/mem/external_slave.hh [hunk 5]
```diff
@@ -123,17 +132,31 @@ class ExternalSlave : public SimObject
...
+     */
     ExternalSlave(ExternalSlaveParams *params);
 
-    /** Port interface.  Responds only to port "port" */
+    /**
+     * Port interface. Responds only to port "port".
+     *
+     * @ingroup api_port
+     */
     Port &getPort(const std::string &if_name,
```
Comment (powerjg) [inline src/mem/external_slave.hh:145]: This should be defined in the sublcass, not here (I think). @BobbyRBruce, do you remember how APIs worked with sublcasses last time?

File: src/mem/port.hh [hunk 6]
```diff
@@ -80,18 +80,27 @@ class MasterPort : public Port, public AtomicRequestProtocol,
     SimObject &owner;
 
   public:
+    /**
+     * @ingroup api_port
+     * @{
+     */
     MasterPort(const std::string& name, SimObject* _owner,
```
Comment (powerjg) [inline src/mem/port.hh:87]: FYI, this is going to conflict with https://gem5-review.googlesource.com/c/public/gem5/+/32308. We want to make sure that is merged first. I don't want to define Master/Slave in the API officially.

File: src/mem/port.hh [hunk 7]
```diff
@@ -287,15 +323,20 @@ class SlavePort : public Port, public AtomicResponseProtocol,
...
      * @return a list of ranges responded to
+     *
+     * @ingroup api_port
      */
     virtual AddrRangeList getAddrRanges() const = 0;
 
     /**
      * We let the master port do the work, so these don't do anything.
+     *
+     * @ingroup api_port
```
Comment (powerjg) [inline src/mem/port.hh:334]: These should be part of the parent class API, not the API here. Again, let Bobby correct me if I'm wrong here.

File: src/mem/ruby/system/RubyPortProxy.hh [hunk 8]
```diff
@@ -58,17 +58,23 @@ class RubyPortProxy : public RubyPort
...
+     *
+     * @ingroup api_port
      */
     virtual ~RubyPortProxy();
 
     /**
      * Initialise a RubyPortProxy by doing nothing and avoid
      * involving the super class.
+     *
+     * @ingroup api_port
```
Comment (powerjg) [inline src/mem/ruby/system/RubyPortProxy.hh:77]: part of SimObj API

File: src/mem/translating_port_proxy.hh [hunk 9]
```diff
@@ -72,18 +72,31 @@ class TranslatingPortProxy : public PortProxy
...
 
-    /** Version of tryReadblob that translates virt->phys and deals
-      * with page boundries. */
+    /**
+     * Version of tryReadblob that translates virt->phys and deals
+     * with page boundries.
+     *
+     * @ingroup api_port
+     */
     bool tryReadBlob(Addr addr, void *p, int size) const override;
```
Comment (powerjg) [inline src/mem/translating_port_proxy.hh:86]: In the parent's API?

**Comment by takekoputa** [comment]
> Generally, I think that RubyPortProxy should not be part of the API, but PortProxy should be.

I saw it being used in learning gem5 part 3: https://gem5.googlesource.com/public/gem5/+/refs/heads/master/configs/learning_gem5/part3/test_caches.py#96

Is there any reason why RubyPortProxy shouldn't be part of the API?

**Comment by powerjg** [comment]
Hmm... good point. Let me mull it over a bit

File: src/mem/external_master.hh (hunk 3 above)
Comment (takekoputa) [reply src/mem/external_master.hh:130]: I find this quite confusing. Even though it's a part of SimObject API, but it has its own signature, i.e. ExternalMasterParams being the parameter.

File: src/mem/external_master.hh (hunk 2 above)
Comment (takekoputa) [reply src/mem/external_master.hh:74]: I agree.

File: src/mem/port.hh (hunk 6 above)
Comment (takekoputa) [reply src/mem/port.hh:87]: Sure, it's gonna be a huge merge conflict. I'd wait until the terminology change to be done before pushing this to gerrit.

File: src/mem/external_master.hh (hunk 3 above)
Comment (powerjg) [reply src/mem/external_master.hh:130]: Yeah... So, the constructors should actually never be used outside of the *Params::create() methods. In some sense, they shouldn't be "public"... In fact, now that I write this, they shouldn't be public! The *Params class should be a friend.

Let's not do make that change now... However, let's not include the constructor as part of the API.

**Comment by powerjg** [comment]
No, RubyPortProxy is just a specific SimObject. There's no need to mark it as an "important" API (same with RubyPort)

File: src/mem/external_master.hh (hunk 2 above)
Comment (takekoputa) [reply src/mem/external_master.hh]: Done

File: src/mem/external_master.hh (hunk 4 above)
Comment (takekoputa) [reply src/mem/external_master.hh]: Removed masterId from the API.

Comment (takekoputa) [reply src/mem/external_master.hh]: anyway, a function in util/tlm/src/sc_master_port.cc  uses masterId to make a request

File: src/mem/external_slave.hh (hunk 5 above)
Comment (takekoputa) [reply src/mem/external_slave.hh:137]: done

File: src/mem/external_master.hh (hunk 1 above)
Comment (takekoputa) [reply src/mem/external_master.hh]: done

File: src/mem/port.hh (hunk 7 above)
Comment (takekoputa) [reply src/mem/port.hh:334]: I'm trying to be consistent here. I assume that bind() and unbind() of Master are also not in the Port API.

File: src/mem/translating_port_proxy.hh (hunk 9 above)
Comment (takekoputa) [reply src/mem/translating_port_proxy.hh:84]: done

File: src/mem/ruby/system/RubyPortProxy.hh (hunk 8 above)
Comment (takekoputa) [reply src/mem/ruby/system/RubyPortProxy.hh]: done

**Comment by takekoputa** [comment]
So should RubyPort a part of the port API? It isn't derived from Port, but it contains 2 subclasses: MemMasterPort and MemSlavePort that are inherited from QueuedMasterPort/QueuedSlavePort.

**Comment by powerjg** [comment]
No, RubyPort shouldn't be part of the PortAPI... it's quite different

File: src/mem/external_master.hh [hunk 10]
```diff
@@ -117,12 +126,15 @@ class ExternalMaster : public SimObject
...
     ExternalMaster(ExternalMasterParams *params);
 
-    /** Port interface.  Responds only to port "port" */
     Port &getPort(const std::string &if_name,
                   PortID idx=InvalidPortID) override;
 
-    /** Register a handler which can provide ports with port_type ==
-     *  handler_name */
+    /**
+     * Register a handler which can provide ports with port_type ==
```
Comment (powerjg) [inline src/mem/external_master.hh:133]: Can you provide any other context here? Do you know how this handler might be used?

**Pull request closed without merging**